### ISSUE TITLE
Integration does not register all received sensors

### ISSUE THREAD

**Issue by gerhardvilsecker** [DominikWrobel/homeassistant-torque#3]
Hello Dominik,
I used original integration from junalmeida for some time. Since it has not been updated for some time and also because i experienced some problems I deinstalled it. As a replacement I found an unfinished torque integration based on ha internal torque code. However this version does take care of making sensors permanent. Since this version ( I installed it as custom_component) is not supported at all, I  installed your fork of junalmeidas bits. It is working in general but: it has same weakpoints as original bits:
1. after restart of HA (or restart of torque_logger) errormessage: "Missing profile name from torque data.". As a consequence no sensor data are updated
2. not all of my configured and in HA received data are  registered as torque sensors
Since I am owner of an electriv vehicle "standard" sensors are not that interesting for me.

I went back to fully unsupported solution thus I cannot provide any logs. But if you want me to do that  I would deactivat current integration and activate yours again. However this is only useful if you are willing to investigate a bit furhter into both issues.
The unsupported integration I am using currently does use torques user-id along with carname to configure integration.
Here is the link to the repository I used to create a custom_component from HA core bits: https://github.com/skyborgff/core/tree/dev/homeassistant/components/torque
Maybe it can provide some ideas for you how to workaround the weaknesses of your current implementation. 
Unfortunately my coding abilitities are very limited. So I cant :-(
PS: I had to change some code in __init__.py of above code in order to get it working, because there were deprecated functions used...
If it would be of help I can provide my custom_component torque setup as tar file...

My setup generally is:
I have a dedicated mobilephone in my car whis is permanently connected to my OBDII adapater, sending permanently telemtry data like:
position, speed, battery charge status, triptime, trip distance
I am using this for generating fully automated driving logbook but also for having history data for electrical consumption of my car.

Regards

Gerhard

**Comment by DominikWrobel** [comment]
First you can try to use the beta version of this integration, It should get more sensors from the ODB interface:

[https://github.com/DominikWrobel/homeassistant-torque/releases/tag/v1.3](url)

I'm also not a programmer, I'm just trying to keep this integration working as long as I can since I use it. If the beta version still is missing some info I will have to find out how to get info from the torque app.

As for the restarts and no data - I don't have this problem, If you can provide logs it would be great.

**Comment by mZer0ne** [comment]
Give me information about what kind of car you have (I'll try to find it), and what obd2 you use, preferably links to some store.
P.S. I also lack some functionality, maybe I'll work on it later!

**Comment by Phoenix260** [comment]
I'm looking for the same thing. Thanks for fixing the code for HA 12.0 :good:

Also not a coder, but if I understand correctly, just need to modify custom_components/torque_logger/const.py to add Custom codes? I tried adding the SOC:

![image_2025-01-02_112915468](https://github.com/user-attachments/assets/a1199903-95a7-4dde-93e1-77cf7658db40)

Although it doesn't seem to be working for me ... and I'm not sure where to put the custom equation that processes the data bits. in this case it should be e/2 for the state of charge (5th bit divide by 2).

I'm using [ODBLink LX](https://www.obdlink.com/products/obdlink-lx/) with a KONA EV (Using the extended PID's from [JejuSoul](https://github.com/JejuSoul/OBD-PIDs-for-HKMC-EVs/blob/master/Hyundai%20Kona%20EV%20%26%20Kia%20Niro%20EV/extendedpids/003_Kona%26Niro_EV_BMS.csv))

**Comment by gerhardvilsecker** [comment]
> Give me information about what kind of car you have (I'll try to find it), and what obd2 you use, preferably links to some store. P.S. I also lack some functionality, maybe I'll work on it later!

Thank you mZerOne: My car is a Cupra born from 2022. I have torque pro, carscanner pro  runinng with MC-Linker OBD2 dongle In addition I have also OBDeleven running for coding. Carscanner is delivereing all the information I would need in realtime. However I have not found a solution so far to send this data to home assistant. Torque is delivering only limited data. However I managed to get a few custom PIDS working. But torque logger does not register these custom pids as sensors in HA. Like Phoenix260 I have also found the const.py file and added my custom pids. But it does not work either. No idea why.

**Comment by DominikWrobel** [comment]
As I understand adding the info to the const.py should add the new sensor to HA, just remember to restart HA after any changes made. If it does not work I will look into this, and maybe @mZer0ne will help.

I made a version that lets you use multiple instances [https://github.com/DominikWrobel/homeassistant-torque/releases/tag/v1.4](url)

Right now I use my car only once a week, so I don't have the time to test this, so any feedback is welcome.

**Comment by gerhardvilsecker** [comment]
Today I found the time to test a little bit. My "findings":
newest version (v1.4) does not work at all in my environment, because Torque is not able to send data to HA (http error message in Torque app)
So I went back to v1.3:
This version works in general. But:
I still have the problem that certain torque pids are not registered in HA. But I can see in debug log that pids are received by torque_logger like:
2025-01-07 18:31:23.143 DEBUG (MainThread) [custom_components.torque_logger] <MultiDictProxy('eml': 'dummy@gmail.com', 'v': '9', 'session': '1736270921039', 'id': '8ac80de713f7c7e63af22ee190fa115b', 'time': '1736271082573', 'kff1005': '8.571475780417703', 'kff1006': '50.12195910004175', 'kff1001': '0.00980858748871849', 'kff1007': '31.104780197143555', 'k222609': '4.5', 'kff1263': '255.0', 'kff1266': '160.382', 'kd': '255.0', 'kff1006': '50.12195910004175', 'kff1239': '24.0', 'kff1010': '178.5605326859697', 'kff1005': '8.571475780417703', 'kff1237': '254.99019141251128', 'k5b': '47.05882352941176', 'k222613': '7.0', 'kff1268': '160.382', 'kff1204': '11.321787499999985')>

k222609 and k222613 are Cupra Born PIDs for outside temperature and inside temperature. As you can see values are correct:
'k222609': '4.5' (outside temp)
'k222613': '7.0 (inside temp)
Both PIDs are ignored by torque_logger
See screenshot:
![image](https://github.com/user-attachments/assets/dedf2f48-6787-49b5-861a-c4b405fbf10c)
I have of course added those pids to your const.py file :-):
  "222609": {
     "shortName": "aussentemp",
     "fullName": "Aussentemperatur",
     "unit": "..C"
  },
  "222613": {
     "shortName": "innentemp",
     "fullName": "Innentemperatur",
     "unit": "..C"
  },

What is really confusing me is the fact that I have also added "HV remaining battery charge" to const.py:
 "5b": {
    "shortName": "hv_battery_charge",
    "fullName": "HV battery reamining Charge",
    "unit": "%"
  },
This is working as you can see from my screenshot. 



Another issue: gps altitude is also provided by torque:  'kff1010': '178.5605326859697
However it is neither reported in device_tracker (as an attribute) nor in sensor part of torque_logger
Of course it is in const.py:
 "ff1010": {
    "shortName": TORQUE_GPS_ALTITUDE,
    "fullName": "GPS Altitude",
    "unit": "m"
  },


Last but not least: "missing profile" problem still exists:
2025-01-07 18:27:44.464 ERROR (MainThread) [custom_components.torque_logger] Missing profile name from torque data.
This happens every time when HA is restarted but Torque not. No sensors are provided as long as you do not restart torque app. After restart of Torque app, "missing profile" error disappears and sensors are available again.
This issue is known for years. To my knowledge it derives from the fact, that torque sends profile data only once at startup. 


Is anyone out there who could fix this issues? I would very appreciate this. Thank you.

**Comment by gerhardvilsecker** [comment]
sorry: forgot to mention that I have restarted HA every time I made changes to const.py

**Comment by gerhardvilsecker** [comment]
Hello all,
I continued testing and troubleshooting. Meanwhile I found workaround for:
a) sensors not appearing in HA
b) Altitude sensor not appearing in HA

Workaround for a)
I dont know why, but torque_logger did not find "unit" for my sensors Aussentemperatur and Innentemperatur. As you can see unit was declared in const.py. Strange. Since I did not find a proper solution I commented out following line of code in coordinator.py:
if (sensor_name and sensor_name != key and
#                len(unit) > 0 and
                key[:3] != ENTITY_GPS and
                key not in self.tracked):
                # do not publish until we have sensor name and unit
                sensor = TorqueSensor(self, self.entry, key, device)
                new_sensors.append(sensor)

len(unit) > 0 asks for an existing unit. Since in my case this field was empty (I added some debug commands in order to see content of variables), I decided to skip this if clause entirely. It worked with some strange side effects:
Sensors where shown in HA but they were somehow greyed out. When I added this to configuration.yaml:
homeassistant:
  customize:
    sensor.mycupra_aussentmp:
      state_class: measurement
      device_class: temperature
      unit_of_measurement: ..C
    sensor.mycupra_innentmp:
      state_class: measurement
      device_class: temperature
      unit_of_measurement: ..C

both sensors showed up correctly after restart

Workaround for b)
I changed exactly the same if clause. But this time I commented out this line:
key[:3] != ENTITY_GPS and

This results in having a device_tracker device plus all single GPS sensors in HA

At the end my if clause looks like this:
if (sensor_name and sensor_name != key and
#                len(unit) > 0 and
#               key[:3] != ENTITY_GPS and
                key not in self.tracked):
                # do not publish until we have sensor name and unit
                sensor = TorqueSensor(self, self.entry, key, device)
                new_sensors.append(sensor)


I dont doubt getting around the "missing profile" error, because to what I learned so far this integration is entirely programmed around getting the torque profile at least for one time. But since Torque does send this profile at startup only we will never survice a HA reboot.

**Comment by gerhardvilsecker** [comment]
typo: I doubt getting around...

**Comment by danielandradam** [comment]
I just wanted to be able to persistently record the data obtained from two different vehicles...
In version 1.3 I can obtain the data from two vehicles without any problems
The problem is that since the PIDs are practically the same, the torque logger somehow replaces the data from one vehicle in the other... in other words, only the telemetry from the last vehicle is stored at the end...
Please help...

**Comment by Failure404** [comment]
> newest version (v1.4) does not work at all in my environment, because Torque is not able to send data to HA (http error message in Torque app) So I went back to v1.3: This version works in general. 

Same, any version newer than 1.3 gives me 404 on ".../api/torque_logger" endpoint.

**Comment by DominikWrobel** [comment]
So I'm trying to add new sensors from the scrach. I've reverted to code to version 1.2 and added new way of adding sensors to HA. As before the new sensors can be defined in const.py. I will be testing this tomorrow, but for now any test will help.

After we can get new sensor to the integration I will work on adding more then one car.

Please try [https://github.com/DominikWrobel/homeassistant-torque/releases/tag/v1.4beta3](url)

**Comment by Insane0923** [comment]
Hi,

Thanks ! This version is OK. Now I don't recieve anymore 404 error.
I'm experiencing a problem and I don't know is only on my car/device, or it's a general problem. The GPS sensor from OBD are not updating instantly on value changes, like the Engine RPM

![Image](https://github.com/user-attachments/assets/6b4f680a-925e-4e7d-91a1-770e7a864285)

I use the official Torque integration and this GPS sensor update instantly on changes.
Now this sensor updated on 10-15 min or even longer :(

![Image](https://github.com/user-attachments/assets/6c52d254-e614-4fd3-b570-3eb7bf145071)

 I really need to relay on these sensors for my use case: Based on GPS sensors of the car OBD, when I enter a set zone, my phone call a number to open the barrier placed at the entrance of my neighborhood

**Comment by danielandradam** [comment]
Greetings to all! Well, I haven't tested the new version yet, because the previous version was already working fine for me. It just wasn't possible to integrate two different sensors (two cars)... However i would like make a comment regarding the location sensors....so that you can test it... in the Torque application, select all the PIDs related to GPS positioning and I think it might work for you... I'm waiting for a new version to integrate more than one vehicle! Hugs

**Comment by DominikWrobel** [comment]
In my setup I have 5 second update interval and the GPS location updates every 5 seconds.

![Image](https://github.com/user-attachments/assets/8a5650ea-3924-4818-9a40-0e0b01edb550)

You can try what @danielandradam wrote and also turn on debugging in the Torque Logger integration and see what data is received by the integration, every message should have GPS tracking in it, like in this example:

```
2025-02-23 19:03:43.436 DEBUG (MainThread) [custom_components.torque_logger] <MultiDictProxy('eml': 'xxxx@xxxxx.xx', 'v': '9', 
'session': '1740329231674', 'id': 'a16a0f7bb05f254a0be0ab3a65b4c4e4', 'time': '1740333822955', 'kff1005': '19.35396043', 
'kff1006': '51.8402194', 'kff1001': '71.60399590587622', 'kff1007': '239.5', 'k33': '100.0', 'kb': '56.0', 'kff1258': '199.5143236688885', 
'kff1267': '160.473', 'kff1268': '2709.237', 'kff1266': '4584.531', 'kff1239': '9.935046195983887', 'kff1005': '19.35396043', 'kff1006': 
'51.8402194', 'kff123b': '239.5', 'kff5202': '14.4490166492815', 'kff5203': '6.920742250302033', 'kff1238': '15.300000190734863', 
'kff1296': '92.727941150848', 'kff1297': '0.7131731080840921', 'kff1298': '6.558885741067913', 'k2f': '31.372549019607842', 'kff126b': 
'31.372549019607842', 'kff126a': '294.65144907836225', 'kff1204': '20.574556967680216', 'kff120c': '4273.379477908607', 'kff1001': 
'71.60399590587622', 'kd': '72.0', 'k46': '3.0', 'kf': '19.0', 'k5': '92.0', 'kff1206': '16.093926797271205', 'kff1010': '205.69195556640625', 
'kff1271': '2.4318120944357307', 'kff125d': '4.008481266535441', 'kff1208': '6.213394733282562')>
```

**Comment by Insane0923** [comment]
I done the following:
1. delete the integration Torque Logger
2. delete the file tracklog.csv on the android phone were I have Torque app installed
3.  reset all the settings on Torque app and reinstall app
4. reinstall Torque Logger integration

Despite the fact I have selected all PIDs related to GPS when I look in HA not a single sensor releted to GPS is recieved.
The only thing that work was to comment out in the "coordinator.py"the line "#key[:3] != ENTITY_GPS and" as @gerhardvilsecker sugested.
But still the sensors are recieved in HA, they don't update :(
I'll try to use the debugger to see if the data is sent to the Torque Logger integration.

I did also a firmware update off the OBD adapter, which is this one https://www.vgatemall.com/products-detail/i-9/

**Comment by stebrau** [comment]
> I'm looking for the same thing. Thanks for fixing the code for HA 12.0 :good:
> 
> Also not a coder, but if I understand correctly, just need to modify custom_components/torque_logger/const.py to add Custom codes? I tried adding the SOC:
> 
> ![image_2025-01-02_112915468](https://github.com/user-attachments/assets/a1199903-95a7-4dde-93e1-77cf7658db40)
> 
> Although it doesn't seem to be working for me ... and I'm not sure where to put the custom equation that processes the data bits. in this case it should be e/2 for the state of charge (5th bit divide by 2).
> 
> I'm using [ODBLink LX](https://www.obdlink.com/products/obdlink-lx/) with a KONA EV (Using the extended PID's from [JejuSoul](https://github.com/JejuSoul/OBD-PIDs-for-HKMC-EVs/blob/master/Hyundai%20Kona%20EV%20%26%20Kia%20Niro%20EV/extendedpids/003_Kona%26Niro_EV_BMS.csv))

10 months have passed - did you find a solution to the data processing? I'm also looking for a solution to the SoC for Kia Soul 27 kWh, with **PID 2101 e/2 for SOC BMS **and** 2105 af/2 for SOC Display.**
I'd appreciate to find a solution.
